cli/azd/internal/telemetry/uploader.go

```diff
@@ -0,0 +1,193 @@
+)
+
+const (
+	maxRetryCount       = 3
+	maxStorageFailCount = 5
+)
+
+var (
+	// Optimistic first retry delay
+	firstTransmitRetryDelay   = time.Duration(100) * time.Millisecond
+	defaultTransmitRetryDelay = time.Duration(2) * time.Second
+	storageQueueRetryDelay    = time.Duration(300) * time.Millisecond
+
+	defaultThrottleDuration = time.Duration(5) * time.Second
+)
+
+type Uploader interface {
+	Upload(ctx context.Context, result chan (error))
+}
+
+type TelemetryUploader struct {
+	transmitter    appinsightsexporter.Transmitter
+	telemetryQueue Queue
+	clock          clock.Clock
+
+	isDebugMode bool
+}
+
+func NewUploader(telemetryQueue Queue, transmitter appinsightsexporter.Transmitter, clock clock.Clock, isDebugMode bool) *TelemetryUploader {
+	return &TelemetryUploader{
+		transmitter:    transmitter,
+		telemetryQueue: telemetryQueue,
+		clock:          clock,
+		isDebugMode:    isDebugMode,
+	}
+}
+
+// Uploads all items that are currently in the telemetry queue.
+// This function returns when no items remain in the queue.
+// An error is only returned if there is a fatal, persistent error with reading the queue.
+func (u *TelemetryUploader) Upload(ctx context.Context, result chan (error)) {
+	for {
+		select {
+		case <-ctx.Done():
+			result <- ctx.Err()
+			return
+		default:
+			done, err := u.uploadNextItem(ctx)
+
+			if done {
+				result <- err
+				return
+			}
+		}
+	}
+}
+
+func (u *TelemetryUploader) uploadNextItem(ctx context.Context) (bool, error) {
+	item, err := u.reliablePeek(ctx)
+
+	if err != nil {
+		return true, fmt.Errorf("failed to peek after %d attempts: %w", maxStorageFailCount, err)
+	}
+
+	if item == nil {
+		return true, nil
+	}
+
+	u.transmit(ctx, item)
+	err = u.reliableRemove(ctx, item)
+
+	if err != nil {
+		return true, fmt.Errorf("failed to remove after %d attempts: %w", maxStorageFailCount, err)
+	}
+
+	return false, nil
+}
+
+func (u *TelemetryUploader) reliablePeek(ctx context.Context) (*StoredItem, error) {
+	var item *StoredItem
+	err := retry.Do(ctx, retry.WithMaxRetries(maxStorageFailCount, retry.NewConstant(storageQueueRetryDelay)), func(ctx context.Context) error {
+		peekItem, err := u.telemetryQueue.Peek()
+
+		if err != nil {
+			return retry.RetryableError(err)
+		}
+
+		item = peekItem
+		return nil
+	})
+
+	if err != nil && ctx.Err() != nil {
+		// Attempt fallback - remove and retry Peek
+		err = u.reliableRemove(ctx, item)
+
+		if err != nil {
+			item, err = u.telemetryQueue.Peek()
+		}
+	}
+
+	return item, err
+}
+
+func (u *TelemetryUploader) reliableRemove(ctx context.Context, item *StoredItem) error {
+	return retry.Do(ctx, retry.WithMaxRetries(maxStorageFailCount, retry.NewConstant(storageQueueRetryDelay)), func(ctx context.Context) error {
+		return retry.RetryableError(u.telemetryQueue.Remove(item))
+	})
+}
+
+func (u *TelemetryUploader) reliableEnqueueWithDelay(ctx context.Context, itemName string, payload []byte, delayDuration time.Duration, attempts int) {
```

Comment:
Should this be returning the error to its caller instead of logging and dropping it?  The other `reliableXYZ` methods seem to follow that pattern.

Reply:
The reason for not propagating `error` similarly is:
If we can't queue an item for retry, and the queue is in a bad state, it's better to proceed with removing the item. (Disk being full for example)

That's the current behavior. `transmit` will end up returning, and the upload loop will remove the item.

I'll add a comment to clarify.